cpp/src/arrow/array/array_test.cc

```diff
@@ -438,6 +439,18 @@ TEST_F(TestArray, TestMakeArrayOfNull) {
       }
     }
   }
+
+  auto req = [](auto type) { return field("", std::move(type), /*nullable=*/false); };
+
+  // union with no nullable fields cannot represent a null
```

Comment:
What would happen with length 0 and length > 1 in the checks below? Perhaps also test that.